### PR TITLE
merge `discardEvents` & `inhibitInput`

### LINKED ISSUE #9043: Brightness control not working from Gestures
KOReader version: 2022.03.1-37
Device: Kindle Paperwhite 5
Firmware version: 5.14.2

Problem: Brightness Control not working from gestures (manual from menu is working good)

Reproduction: adding this functions to Gesture Manager:

Gesture: Swipe Up (and Down)

**Screen and lights:**
* Set Frontlight brightness to 15 (7)
* Set Frontlight warmth to 0 (100)
* Set Night mode Off (On)

**Reflowable document:**
* Set Font size to 20 (28)

Warmth, Night Mode and Font Size control is working good

The Frontlight brightness control does not work, neither up nor down!

### PR DESCRIPTION
**Pull request by yparitcher**
Fixes #9043

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9065)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Yup, looks pretty much like what I imagined ;).

(It's 3AM, though, so, err take that with a grain of salt :D).

I'd possibly unschedule any pending tasks on `false`/`nil`, but that's mostly harmless unless you hit some weird chains on `discardEvents` calls with unfortunate timings.

**Comment by yparitcher** [comment]
@NiLuJe 
Is `inhibitInput` reentrant? Can i call it with (true) multiple time before i call it with (false)

**Comment by yparitcher** [comment]
The call sites for `setIgnoreInput` seem not to want to disable input but do it anyways on android so as not to ANR, so not sure if should be merged.

**Comment by NiLuJe** [comment]
> @NiLuJe 
> Is `inhibitInput` reentrant? Can i call it with (true) multiple time before i call it with (false)

Yep!

**Comment by NiLuJe** [comment]
> The call sites for `setIgnoreInput` seem not to want to disable input but do it anyways on android so as not to ANR, so not sure if should be merged.

Yeah, at worst this makes the inhibit stuff slightly redundant, but at least it ensures we'll drain whatever left in the event queue without actually triggering anything on our end.

File: frontend/device/input.lua [hunk 1]
```diff
@@ -1334,4 +1334,37 @@ function Input:inhibitInput(toggle)
...
+        else
+            -- On non-eInk screen, display is usually instantaneous
+            delay = 0.4
+        end
+    else -- we expect a number
+        delay = set_or_seconds
+    end
+    local UIManager = require("ui/uimanager")
+    UIManager:scheduleIn(delay, function() self:inhibitInputUntil() end)
+    self:inhibitInput(true)
```
Comment (poire-z) [inline frontend/device/input.lua:1367]: Feels like a risk if a second inhibitInputUntil() is issued before a first one is done (dunno if that can happen in our code) - the first one would have this scheduleIn'ed uninhibit cancel the 2nd one.
(Previously, with the discard_events_till, further calls would just increase its value.)

Comment (yparitcher) [reply frontend/device/input.lua]: Done

**Comment by yparitcher** [comment]
Rebased after #8999

File: frontend/device/input.lua [hunk 2]
```diff
@@ -1333,4 +1336,38 @@ function Input:inhibitInput(toggle)
...
+@param set_or_seconds either `true`, in which case a platform-specific delay is chosen, or a duration in seconds (***int***).
+]]
+function Input:inhibitInputUntil(set_or_seconds)
+    local UIManager = require("ui/uimanager")
+    UIManager:unschedule(self._inhibitInputUntil_func)
+    if not set_or_seconds then -- remove any previously set
+        self:inhibitInput(false)
+        return
+    end
+    local delay
```
Comment (zwim) [inline frontend/device/input.lua:1351]: Please change that to delay_s

File: frontend/device/input.lua [hunk 3]
```diff
@@ -1333,4 +1336,38 @@ function Input:inhibitInput(toggle)
...
+    if set_or_seconds == true then
+        -- Use an adequate delay to account for device refresh duration
+        -- so any events happening in this delay (ie. before a widget
+        -- is really painted on screen) are discarded.
+        if self.device:hasEinkScreen() then
+            -- A screen refresh can take a few 100ms,
+            -- sometimes > 500ms on some devices/temperatures.
+            -- So, block for 400ms (to have it displayed) + 400ms
+            -- for user reaction to it
+            delay = 0.8
```
Comment (zwim) [inline frontend/device/input.lua:1361]: ditto

**Review by Frenzie — APPROVED** [review]
lgtm

File: frontend/device/input.lua [hunk 4]
```diff
@@ -1333,4 +1336,38 @@ function Input:inhibitInput(toggle)
...
+        else
+            -- On non-eInk screen, display is usually instantaneous
+            delay = 0.4
```
Comment (zwim) [inline frontend/device/input.lua:1364]: ditto

File: frontend/device/input.lua [hunk 5]
```diff
@@ -1333,4 +1336,38 @@ function Input:inhibitInput(toggle)
...
+        end
+    else -- we expect a number
+        delay = set_or_seconds
+    end
+    UIManager:scheduleIn(delay, self._inhibitInputUntil_func)
```
Comment (zwim) [inline frontend/device/input.lua:1369]: ditto

**Comment by zwim** [comment]
@yparitcher  Now #8999 is merged and `timeval` is not working well anymore. Esp. `UIManager:gettime` and Co. are now using a `time` object (a number).

I want to give you a short tutorial (details are in the doc of `time.lua`), bring you up to date.

Variable namings: 
* If we use a variable name that implies a time (duration, now, delay ...) then it should use the new `time.xxx` functions. 
* If we use a variable containing a time with a certain unit we append that (e.g. delay_s, duration_m ...)

**Review by NiLuJe — APPROVED** [review]
Good call on moving everything to `Input` (and the rename ;)).

File: frontend/device/input.lua (hunk 2 above)
Comment (yparitcher) [reply frontend/device/input.lua]: Done

**Comment by yparitcher** [comment]
Ready

**Comment by poire-z** [comment]
Mhhh, I updated today (haven't for a few weeks), and I got a few issues with my own patch that may be related to this change, but that I have a hard time figuring out... @NiLuJe : do you have any idea what might be happening ?

I have something a bit like @zwim is doing at https://github.com/koreader/koreader/pull/9028#issuecomment-1137047170 (dunno if you're having the same issue I have if you try to use inhibitInputUntil ?), that is:
- I'm in some pre-auto-suspend state where I set the frontlight off - and then I wait 30 seconds for some event and if none, I go into suspend.
- if the user tap, I get first a "touch" event, that is received, processed by some of my code in uimanager. Before, I did UIManager:discardEvent(true), now I do Input:inhibitInputUntil(true) in the hope that the "release" event from my tap is eaten and discarded - which worked previously, but now get a strange behaviour: it becomes a hold !

Some log:

```
05/26/22-19:37:39 DEBUG CreDocument: goto page 343 flow 0
05/26/22-19:37:46 INFO  task @ frontend/ui/uimanager.lua 1931 0 (this is my multipupose schedule_action)
05/26/22-19:38:09 INFO  task @ frontend/ui/uimanager.lua 1931 0
frontglight off, real suspend schedule for in 30s

I tap on the screen
05/26/22-19:38:15 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 1, time: 3976.382089
05/26/22-19:38:15 DEBUG input event => type: 3 (EV_ABS), code: 48 (ABS_MT_TOUCH_MAJOR), value: 1, time: 3976.382115
05/26/22-19:38:15 DEBUG input event => type: 3 (EV_ABS), code: 50 (ABS_MT_WIDTH_MAJOR), value: 1, time: 3976.382117
05/26/22-19:38:15 DEBUG input event => type: 3 (EV_ABS), code: 53 (ABS_MT_POSITION_X), value: 1071, time: 3976.382122
05/26/22-19:38:15 DEBUG input event => type: 3 (EV_ABS), code: 54 (ABS_MT_POSITION_Y), value: 218, time: 3976.382126
05/26/22-19:38:15 DEBUG input event => type: 3 (EV_ABS), code: 24 (ABS_PRESSURE), value: 1024, time: 3976.382132
05/26/22-19:38:15 DEBUG key event => code: 330 (BTN_TOUCH), value: 1, time: 3976.382148
05/26/22-19:38:15 DEBUG input event => type: 0 (EV_SYN), code: 2 (SYN_MT_REPORT), value: 0, time: 3976.382151
05/26/22-19:38:15 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 3976.382167
05/26/22-19:38:15 DEBUG in tap state...
05/26/22-19:38:15 DEBUG set up hold timer
05/26/22-19:38:15 DEBUG adjusted ges: touch
05/26/22-19:38:16 INFO  Inhibiting user input
this should inhibit future events for 0.8s, but:

05/26/22-19:38:16 INFO  task @ frontend/ui/uimanager.lua 1931 1
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 1, time: 3976.416936
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 48 (ABS_MT_TOUCH_MAJOR), value: 1, time: 3976.416955
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 50 (ABS_MT_WIDTH_MAJOR), value: 1, time: 3976.416957
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 53 (ABS_MT_POSITION_X), value: 1071, time: 3976.416970
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 54 (ABS_MT_POSITION_Y), value: 221, time: 3976.416974
05/26/22-19:38:16 DEBUG input event => type: 0 (EV_SYN), code: 2 (SYN_MT_REPORT), value: 0, time: 3976.417003
05/26/22-19:38:16 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 3976.417016
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 1, time: 3976.425942
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 48 (ABS_MT_TOUCH_MAJOR), value: 0, time: 3976.425972
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 50 (ABS_MT_WIDTH_MAJOR), value: 0, time: 3976.425974
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 53 (ABS_MT_POSITION_X), value: 1071, time: 3976.425977
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 54 (ABS_MT_POSITION_Y), value: 221, time: 3976.425980
05/26/22-19:38:16 DEBUG input event => type: 3 (EV_ABS), code: 24 (ABS_PRESSURE), value: 0, time: 3976.425984
05/26/22-19:38:16 DEBUG key event => code: 330 (BTN_TOUCH), value: 0, time: 3976.425999
05/26/22-19:38:16 DEBUG input event => type: 0 (EV_SYN), code: 2 (SYN_MT_REPORT), value: 0, time: 3976.426001
05/26/22-19:38:16 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 3976.426016
05/26/22-19:38:16 DEBUG hold gesture detected in slot 1
05/26/22-19:38:16 DEBUG in hold state...
05/26/22-19:38:16 DEBUG adjusted ges: hold
but I got an event, a hold ! which generates some text selection start:

05/26/22-19:38:16 DEBUG clear selection
05/26/22-19:38:16 DEBUG document has no pages at {
    ["y"] = 1071,
    ["w"] = 0,
    ["page"] = 343,
    ["x"] = 854,
    ["h"] = 0,
}
05/26/22-19:38:16 DEBUG hold position in page {
    ["y"] = 1071,
    ["w"] = 0,
    ["page"] = 343,
    ["x"] = 854,
    ["h"] = 0,
}
05/26/22-19:38:16 DEBUG CreDocument: get text range {
    ["pos0"] = "/body/DocFragment[23]/body/p[75]/text()[2].464",
    ["pos1"] = "/body/DocFragment[23]/body/p[75]/text()[2].468",
    ["text"] = "vrai",
}
05/26/22-19:38:16 DEBUG selected word: {
    ["page"] = 343,
    ["pos0"] = "/body/DocFragment[23]/body/p[75]/text()[2].464",
    ["pos1"] = "/body/DocFragment[23]/body/p[75]/text()[2].468",
    ["word"] = "vrai",
    ["sbox"] = {
        ["y"] = 1025,
        ["w"] = 65,
        ["h"] = 49,
        ["x"] = 854,
    },
}
05/26/22-19:38:16 DEBUG getLinkFromPosition link_xpointer:
05/26/22-19:38:16 DEBUG getLinkFromPosition a_xpointer:
05/26/22-19:38:16 DEBUG _refresh: Enqueued ui update for region 854 1025 65 49
05/26/22-19:38:16 DEBUG setDirty ui from widget ReaderUI w/ region 854 1025 65 49
05/26/22-19:38:16 DEBUG painting widget: ReaderUI
05/26/22-19:38:16 DEBUG CreDocument: goto page 343 flow 0

and now only is user input un'inhibited:
05/26/22-19:38:16 INFO  task @ frontend/device/input.lua 263 1
05/26/22-19:38:16 INFO  Restoring user input handling

and next event works (well, my tap is considered a hold release, but that's because of the strange state happening above)
05/26/22-19:38:26 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 1, time: 3986.800631
```

Here's the log for a classic tap when nothing is inhibited:
```
05/26/22-19:37:33 DEBUG _refresh: Enqueued ui update for region 0 0 1072 1086
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 1, time: 3935.784202
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 48 (ABS_MT_TOUCH_MAJOR), value: 1, time: 3935.784221
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 50 (ABS_MT_WIDTH_MAJOR), value: 1, time: 3935.784223
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 53 (ABS_MT_POSITION_X), value: 1262, time: 3935.784234
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 54 (ABS_MT_POSITION_Y), value: 727, time: 3935.784237
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 24 (ABS_PRESSURE), value: 1024, time: 3935.784248
05/26/22-19:37:35 DEBUG key event => code: 330 (BTN_TOUCH), value: 1, time: 3935.784265
05/26/22-19:37:35 DEBUG input event => type: 0 (EV_SYN), code: 2 (SYN_MT_REPORT), value: 0, time: 3935.784269
05/26/22-19:37:35 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 3935.784282
05/26/22-19:37:35 DEBUG in tap state...
05/26/22-19:37:35 DEBUG set up hold timer
05/26/22-19:37:35 DEBUG adjusted ges: touch
got my touch
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 1, time: 3935.829692
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 48 (ABS_MT_TOUCH_MAJOR), value: 0, time: 3935.829724
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 50 (ABS_MT_WIDTH_MAJOR), value: 0, time: 3935.829726
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 53 (ABS_MT_POSITION_X), value: 1262, time: 3935.829729
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 54 (ABS_MT_POSITION_Y), value: 727, time: 3935.829735
05/26/22-19:37:35 DEBUG input event => type: 3 (EV_ABS), code: 24 (ABS_PRESSURE), value: 0, time: 3935.829748
05/26/22-19:37:35 DEBUG key event => code: 330 (BTN_TOUCH), value: 0, time: 3935.829764
05/26/22-19:37:35 DEBUG input event => type: 0 (EV_SYN), code: 2 (SYN_MT_REPORT), value: 0, time: 3935.829766
05/26/22-19:37:35 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 3935.829783
05/26/22-19:37:35 DEBUG in tap state...
05/26/22-19:37:35 DEBUG single tap detected in slot 1 {
    ["y"] = 1262,
    ["w"] = 0,
    ["h"] = 0,
    ["x"] = 345,
}
05/26/22-19:37:35 DEBUG adjusted ges: tap
got my tap
05/26/22-19:37:35 DEBUG close widget: table: 0x2cf88ef8
```
the events' time and time difference/delays don't seen that different, and smaller enough from the "long tap interval" value I haven't changed from 500ms (HOLD_INTERVAL_MS = 500).
Any idea what might be happening ?

I also get the feeling that some of the dictionary popup quick close/prevent close if tap just when it's about to be shown - are not working as they did before (but I'm not fresh about how that's supposed to work :) so I'm not really sure, will have to investigate more...

**Comment by yparitcher** [comment]
I suspect that the major difference between `discardEvents` & `inhibitInput` is that:
`discardEvents` discards complete input events after being processed by `Input` so I think that the complete touch event (down & up) is discarded (It does no know from press or release).
`inhibitInput` gets rid if the `Input` handler for these events meaning the kernel (i think) event will be dropped, therefore if started in middle of a touch event, it probably just eats the release causing it to end up being a hold.

Discarding the events is not the correct solution, however the inhibit input should be a little smarter and detect if we are in middle of input

**Comment by zwim** [comment]
For me the trapwidget works perfectly. It gets closed either per tap or per UIManager:close().

Unfortunately I have not time until Wdnesday to finish the other PR as I am hunting real bugs now in an Tyrolean national park now. :)

**Comment by yparitcher** [comment]
specifically this line i suspect looks for uncompleted touches and makes them a hold. @NiLuJe ?

**Comment by NiLuJe** [comment]
Might need a state reset or something. I'll look into it on Saturday ;).

**Comment by NiLuJe** [comment]
By which I mean `Input:resetState()` like what happens on Suspend/Resume (can't remember which one), to avoid just such issues ;).

**Pull request merged**